workflows/cloudify_system_workflows/snapshot.py

```diff
@@ -0,0 +1,574 @@
+        try:
+            while s:
+                obj, idx = decoder.raw_decode(s)
+                n += 1
+                yield json.dumps(obj)
+                s = s[idx:]
+        except:
+            pass
+
+    assert not n or not s
+
+
+def _copy_data(archive_root, config, to_archive=True):
+    DATA_TO_COPY = [
+        (config.file_server_blueprints_folder, 'blueprints'),
+        (config.file_server_uploaded_blueprints_folder, 'uploaded-blueprints')
+    ]
+
+    # files with constant relative/absolute paths
+    for (p1, p2) in DATA_TO_COPY:
+        if p1[0] != '/':
+            p1 = os.path.join(config.file_server_root, p1)
+        if p2[0] != '/':
+            p2 = os.path.join(archive_root, p2)
+        if not to_archive:
+            p1, p2 = p2, p1
+
+        if not os.path.exists(p1):
+            continue
+
+        if os.path.isfile(p1):
+            shutil.copy(p1, p2)
+        else:
+            if os.path.exists(p2):
+                shutil.rmtree(p2, ignore_errors=True)
+            shutil.copytree(p1, p2)
+
+
+def _create_es_client(config):
+    return elasticsearch.Elasticsearch(hosts=[{'host': config.db_address,
+                                               'port': int(config.db_port)}])
+
+
+def _except_types(s, *args):
+    return (e for e in s if e['_type'] not in args)
+
+
+def _clean_up_db_before_restore(es_client, wf_exec_id):
+    s = elasticsearch.helpers.scan(es_client)
+    for doc in _except_types(s, 'provider_context', 'snapshot'):
+        if doc['_id'] != wf_exec_id:
+            doc['_op_type'] = 'delete'
+            yield doc
+
+
+def _create(ctx, snapshot_id, config, include_metrics, include_credentials,
+            **kw):
+    tempdir = tempfile.mkdtemp('-snapshot-data')
+
+    snapshots_dir = os.path.join(
+        config.file_server_root,
+        config.file_server_snapshots_folder
+    )
+
+    if not os.path.exists(snapshots_dir):
+        os.makedirs(snapshots_dir)
+
+    metadata = {}
+
+    # files/dirs copy
+    _copy_data(tempdir, config)
+
+    # elasticsearch
+    es = _create_es_client(config)
+    has_cloudify_events = es.indices.exists(index=_EVENTS_INDEX_NAME)
+    _dump_elasticsearch(tempdir, es, ctx.execution_id, has_cloudify_events)
+
+    metadata[_M_HAS_CLOUDIFY_EVENTS] = has_cloudify_events
+
+    # influxdb
+    if include_metrics:
+        _dump_influxdb(tempdir)
+
+    # credentials
+    if include_credentials:
+        ctx.load_deployments_contexts()
+        _dump_credentials(ctx, tempdir)
+
+    # version
+    metadata[_M_VERSION] = _VERSION
+
+    # metadata
+    with open(os.path.join(tempdir, _METADATA_FILE), 'w') as f:
+        json.dump(metadata, f)
+
+    # agents
+    _dump_agents(tempdir, ctx)
+
+    # zip
+    snapshot_dir = os.path.join(snapshots_dir, snapshot_id)
+    os.makedirs(snapshot_dir)
+
+    shutil.make_archive(
+        os.path.join(snapshot_dir, snapshot_id),
+        'zip',
+        tempdir
+    )
+
+    # end
+    shutil.rmtree(tempdir)
```

Comment:
Perhaps use the try..finally pattern here for making sure everything gets cleaned up in a case of failure.